src/toil/job.py

```diff
@@ -3026,3 +3031,22 @@ def convertPromises(kwargs):
                 foundPromisedRequirement = True
         return foundPromisedRequirement
 
+
+class JobStoreFile(str):
+    """
+    A class to wrap the job store file id returned by writeGlobalFile and any attributes we may want
+    to add to it.
+    """
+    # Need to modify this so this subclass of the mutable str can accept two arguments.
+    @classmethod
+    def __new__(cls, *args):
+        # args is (<class>, fileStoreID, filePath) during instantiation, and (<class>, fileStoreID)
+        # during unpickling.  We care about the fileStoreID.
+        return str.__new__(cls, args[1])
+
+    def __init__(self, fileStoreID, filePath):
```

Comment:
The constructor should have no intelligence, i.e. be dumb. That will give you more freedom to construct objects of this class in different ways. You can tell that there is sth wrong with this approach because it smells: the bad smell is that _resolveFileSize sets an instance attribute. PyCharm actually warns you about that.

The signature of the constructor should be `__init__(self, fileID, size)`. Then add a class method that has this signature: `forPath(cls, fileID, path)`. Construction then reads like FileInfo.forPath().